### DISCUSSION TITLE
RFC: Continuous Tasks

### DISCUSSION THREAD

**Discussion by FrozenPandaz** [nrwl/nx#29025] [Feature Requests]
# RFC: Continuous Tasks

Some tasks run through Nx are tasks which run continuously until they are terminated. In this RFC, these tasks will be referred to as continuous tasks. Examples of continuous tasks are tasks which start servers, build and test tasks which have watch modes, and many more. Currently, Nx expects all tasks to end and does not really have much special handling for continuous tasks. In many cases, tasks have a relationship with these continuous tasks where they depend on them being running. If this is specified as a `dependsOn` today, Nx will never continue as it will wait for the depended upon continuous tasks to finish; which it will not.

## Concept

Nx should have handling for continuous tasks. Continuous tasks are a subset of tasks which never end on their own accord; they must be terminated by something else.

Continuous tasks are much like other tasks. Continuous tasks can be run like any other tasks, either on their own, as part of a set of tasks, or as a dependency of another task. Continuous tasks can depend on other tasks; continuous and discrete. And other tasks, continuous or not, can depend on continuous tasks.

Continuous tasks are different from discrete tasks in some ways though. Continuous tasks do not end on their own and need to be terminated by something else. Continuous tasks also do not yield an output; this means that they cannot be cached. However, continuous tasks do yield a side effect which is likely the purpose why the task is kept alive.

Nx will add the concept of Continuous tasks which Nx will orchestrate as part of the task graph. This will allow for tasks to depend on other _running_ tasks in addition how it already allows depending on other _completed_ tasks. 

### Starting a web-server to e2e test

Currently in Nx, the e2e task is responsible for starting the web-server which it is testing. With dependencies on continuous tasks, Nx can takeover the responsibility of starting the web server while the e2e task is responsible for waiting for the server to be ready. Cypress, Playwright and other e2e tasks runners already do this.

### Starting backends alongside a web server

```js
{
  name: 'frontend',
  targets: {
    serve: {
      command: 'vite dev',
      dependsOn: ['content-backend:serve', 'auth-backend:serve']
    }
  }
}
```

This would start the required backends while running a frontend application.

### Starting a db alongside a backend server

```js
{
  name: 'api',
  targets: {
    serve: {
      command: 'node ./server.js',
      dependsOn: ['database:start']
    }
  }
}
```

### Publishing to a local-registry before running e2e tests

```js
{
  name: 'angular-e2e',
  targets: {
    serve: {
      command: 'jest e2e/angular',
      dependsOn: ['@nx/source:prepare-local-registry']
    }
  }
}
```

## Feedback Requested

While you read this RFC and dive into the details, the Nx core team would like some feedback on this design before we implement it. Please let us know your thoughts after you have read it.

### Is this a feature you would like to see implemented?

Firstly, is this a feature that excites you and does it solve the problems you face? And would this feature improve or enable the kinds of workflows that you need?

### Naming/ Terminology

After this discussion, it seems like continuous tasks is a good term to move forward with. If you have any objections to this, please continue to discuss.

Secondly, naming things is hard. "continuous tasks" is terminology which is subject to change. We would like to hear feedback on the terminology we should use. We'd like the term to be easily understood and intuitively capture this type of task and the differences with the existing concept of tasks.

Infinite Tasks was a considered name. Infinite refers to the infinite duration of the tasks but other terms could be more specific. Is it obvious what infinite vs finite pertains to?

Perpetual could make sense because the tasks remain running as does a perpetual motion machine. Unfortunately, there is no real antonym as there is for infinite vs finite. Talking about not perpetual tasks is not as nice.

We have also considered some other terms.

Tasks can be classified as neverending or endless as they will not end on their own accord. They would end though, when they are terminated.

These tasks could be considered long running tasks. But some tasks such as e2e tests can also run for a long time while not being continuously running. And if these tasks don't last for that long... is it fair to call them long running?

Immortal vs Mortal is another way of describing these tasks. Immortallity is specific to life though but tasks aren't living. It also has religious connotations which may cause uncomfort.

Tasks should eventually complete. If something does not complete, is it even still a task? Should we introduce a new term altogether? Rather than referring to them as tasks, should we refer to them as processes? This would be easily confused for OS processes. Comment below if you think it's confusing for both of these concepts to be referred to as tasks.

### Timeline

We hope to begin implementation starting January 2025 so please provide as much feedback as you can before then.

## Lifecycle

The lifecycle for a continuous task would have the following lifecycle:
1. Waits for dependent discrete tasks to complete and dependent continuous tasks to start
2. Task starts
3. Task is running
4. Task is terminated

Now we'll contrast this to the lifecycle for a discrete task:
1. Waits for dependent discrete tasks to complete and dependent continuous tasks to start
2. Task starts
3. Task is running
4. Task completes either successfully or not

There are 2 main differences from what is currently in Nx.

Tasks (both discrete and continuous) depending on continuous tasks, Nx will not wait for the task to complete. It will instead, wait for the task to start. The task may not be fully ready to produce side effects yet. This will be discussed later on.

Continuous tasks need to be terminated. Nx would handle this but when this occurs depends on why Nx is running that continuous task.

## Use Cases

Below, we'll dive into some different cases Nx will have to handle. Each of them will detail some current workarounds and their flaws. @vsavkin also produced a video with diagrams to go over this topic here: https://www.youtube.com/watch?v=-gezeX9zxuM

### Isolated Continuous Task

Running a continuous task on its own with no dependencies and without other tasks depending on it is already fairly well supported by Nx. If you're using Nx to run `app:serve`, `app:build --watch`, or `app:test --watch` then these are already continuous tasks. From the user's point of view, this should be the exact same as if the user were to run that task outside of Nx.

When a continuous task starts running, it is expected to continue running until Nx ends it. Thus, Nx would throw an error if a continuous task were to exit before Nx terminates it. This will be true for the following cases as well.

Nx starts the continuous task solely because the user instructed it to. There is not much other known information as there will be in other use cases. Nx will thus wait for the user to interrupt Nx before it terminates the continuous task.

### Continuous Tasks Depending on Discrete Tasks

Continuous tasks can depend on discrete tasks.

For example, `app:serve` could depend on itself and it's dependencies to be built before starting up.

Running a continuous task on its own with dependencies on other tasks but no dependencies on itself will be much like running a continuous task in isolation. The only difference is that Nx will wait for the discrete tasks to complete before the continuous task is started. It is expected to continue running until Nx is interrupted (Ctrl + C).

### Multiple Unrelated Continuous Tasks

Running multiple continuous tasks may be something users already do.

An example of this is serving two applications via [`nx run-many -t serve --projects frontend,backend`](https://nx.dev/nx-api/nx/documents/run-many#runmany) such as a frontend and a backend.

As it already does currently, Nx will wait dependent tasks of each continuous task before starting them respectively. As with the prior cases, Nx would also terminate these tasks when the user interrupts Nx.

The main change here will be how Nx handle the terminal outputs. Currently, the user has a few different options for how to show the output. Only 2 of the 3 really make sense.
* Static
    * This mode would display outputs of dependent tasks when they are done but then would not display the outputs of continuous tasks since they do not end. This option doesn't make much sense.
* Stream
    * This mode interlaces the terminal outputs while prefixing them. This is alright but output is still harder to understand as it can be broken up.
* Dynamic (default) and "stream-without-prefixes" which is similar
    * This mode interlaces the terminal outputs without prefixing them.

Nx would be updated to handle multiple streams of outputs differently. What would be ideal in this case is if the user could switch focus between different continuous task outputs. None of the continuous tasks are auxilliary and all of them are main tasks. Terminal output will be covered further in more detail below.

### Single Discrete Task Depending on Continuous Task(s)

A discrete task can depend on continuous task(s).

For example, e2e tests can depend on the application server. Nx would start the application server first, then start the e2e tests. Until the e2e tests complete, the application server will stay running. When the e2e tests finish, Nx will terminate the application server.

Currently, Nx does not handle this case. Because tasks wait for all dependent tasks to end and continuous tasks do not end, Nx will just stop at running the continuous task.

Nx plugins can do their best to handle this case by configuring tools to wait for another Nx task to start and produce side effects. Doing this does handle the same issue but is inefficient as it will yield multiple task graphs as opposed to being combined in a single task graph. This would mean that tasks which exist in multiple graphs, are run once for each graph being executed. Caching can be utilized to reuse as much existing work as possible. Not all work can be cached though so this could result in a lot of wasted work.

When a discrete task depends on any continuous task(s), Nx will start the continuous task(s) before running the discrete task. Unlike the previous cases, the user starting the continuous task directly. The continuous task is started to accompany the discrete task. The discrete task will eventually complete. When this happens, the continuous task(s) become orphaned and Nx will terminate them since they are no longer necessary.

As for the actual nature of the dependency, the discrete task depending on the continuous task(s) likely needs some sort of side effect to be produced before the discrete task can truly begin. If Nx were to start the discrete task immediately as soon as the continuous task is started, the discrete task may encounter issues if it does not handle the side effect not being present. For the initial implementation, Nx will classify that as an issue with the discrete task. When the tasks are run outside of Nx, they still likely need some handling for when the side effect does not exist. Application servers need to handle when their accompanying services are unavailable. And test runners are already currently configured to wait for an application server to come up. However, Nx can still prioritize starting the continuous task(s) being depended upon as soon as it can. This would give more time for the side-effect to become available while Nx runs other dependent tasks. But ultimately, Nx will not further delay the execution of tasks depending on continuous task(s) after the continuous task has started.

The terminal output of the discrete task is more important than the output of the continuous task(s) output in this case. The user initiated directly. Nx would have the discrete task's terminal output shown most prominently. The continuous task(s) output could either be hidden entirely or optionally shown. Ideally, Nx would allow the user to see two terminal outputs at once. Again, terminal output will be covered in more detail further below.

### Multiple Discrete Tasks depending on a particular Continuous Task(s)

An extension of the previous case is when multiple discrete tasks depend on a single continuous task or set of continuous tasks.

An example of this is multiple e2e test suites (`app1-e2e:e2e` and `app2-e2e:e2e` both depending `app:serve`) or atomized test suites (`app1-e2e:e2e-ci--feature1.spec.ts` and `app1-e2e:e2e-ci--feature2.spec.ts` both depending on `app:serve`).

As with the previous case, Nx Plugins do their best to handle this case by configuring the testing tool to start the continuous task. This has the same issues as the previous case but it is amplified because each and every task depending on the continuous task would waste some work. A lot of compute is wasted here.

When multiple discrete tasks depend on a particular continuous task or a set of continuous tasks, Nx can reuse the continuous task(s). As with the previous case, Nx will start the continuous task(s) being depended upon as soon as possible. However, Nx will not terminate the continuous task(s) until all tasks which depend upon the continuous task completes.

This case is more complicated to distribute across different machines. Multiple discrete tasks would be in the execution queue. An agent does not know which if any of those discrete tasks it will receive. Agents could prematurely run any continuous tasks any task in the execution queue depends on but this may waste some work if some agents end up not being assigned any tasks depending on those continuous tasks. Agents could only run the continuous task necessary for tasks it is assigned when it is assigned but this would likely leave the tasks depending on the continuous tasks waiting for a bit until the side effects are available. Ideally, the orchestrator would accurately make the decision upfront about which subset of agents will receive the dependent tasks and run them efficiently utilizing only those agents. There's a lot of optimization which can be done but to start, it should be sufficient for the orchestrator to send continuous task(s) to an agent before it wants to send it a task which depends on it. We'll start there and explore more optimizations later on.

As for the terminal output, now multiple discrete tasks would need to have their outputs shown while also showing the output of continuous tasks at the same time. Nx already handles showing multiple discrete tasks' output, but not the output of dependent continuous task's. The mortal task output should still be more prominent and the dependent tasks could be optionally shown. Again, terminal output will be covered in more detail further below.

## Continuous Task(s) Depending on Continuous Task(s)

An extension of all the previous cases, is when continuous task(s) depend on a particular continuous task or a set of continuous tasks.

Examples of this would be if:
* `frontend:serve` depends on `backend:serve`
* `frontend:serve` depends on `backend1:serve` and `backend2:serve`
* `app1:serve` and `app2:serve` both depend on `backend:serve`
* `app2:serve` and `app2:serve` both depend on both `backend1:serve` and `backend2:serve`
* `backend:serve` depends on `database:serve`
* `app1:serve` and `app2:serve` both depend on `backend1:serve` and `backend2:serve` which depend on `database1:serve` and `database2:serve` respectively

Currently, Nx does not handle this because the continuous tasks do not end and Nx will not continue.

Nx plugins again can do their best to handle this case using an executor to optionally start the dependent servers only of they are not available. This has a lot of intricacies. When multiple tasks are started simultaneously, there is still a chance multiple tasks will start the dependent servers as they are not available. There's a lot of wasted work as it's not included in the task graph. In theory, this can be handled perfectly.. but honestly, this case is not handled well for the most part at the moment.

Nx could handle this case with the same basic principles discussed in the previous sections. Dependent Continuous tasks are started first before any continuous tasks depending on them are started. The multiple continuous tasks started directly by the user remain running until the user interrupts Nx. And when this happens, the dependent continuous tasks are also terminated.

### Configuration

A target will declare that the task that it creates is a continuous task.

```js
{
  name: 'app',
  targets: {
    serve: {
      command: 'vite',
      continuous: true
    }
  }
}
```

Other targets can specify that its task depend on the continuous task

```js
{
  name: 'app-e2e',
  targets: {
    e2e: {
      command: 'playwright test',
      dependsOn: ['app:serve']
    }
  }
}
```

With this configuration, Nx will know that `app:serve` is infinite.

### Readiness

Earlier we touched on the fact that tasks can take some time to produce side effects which parent tasks are expecting. For instance, starting servers might not be instantaneous. Continuous tasks may need a way to indicate that the side effects have been produced. Parent tasks would then wait for their dependent tasks to be ready rather than started. This concept of "readiness" does not currently exist.

#### Option 1: Do not handle it

Tasks are likely already built in a way where they can be started without their dependent continuous tasks being truly ready and wait until the side effects are available. It is definitely an option to not introduce the notion of "ready" to Nx and let the task itself handle this. Given this, Nx would start the parent task possibly immediately after the dependent continuous tasks have started.

Applications would need this sort of logic when running in production (not via Nx), so most applications should already be built to handle this kind of startup. For example, frontends would need to handle if their backends either haven't started or went down. Not delaying parent tasks allows multiple continuous tasks begin any pre work at the same time.

For tests, the plugins that Nx provides for Cypress and Playwright already handle waiting for the web server to startup. This waiting would continue to exist without Cypress and Playwright explicitly responsible for starting and terminating the dependent continuous tasks.

If the parent task doesn't handle waiting for the side effect, the parent task would fail and Nx would leave it up to developers to change the behavior of the parent task.

#### Option 2: Add configuration

Tasks can report to Nx that the sidecar task produces a side effect. Nx can handle waiting for these side effects. Doing so would delay parent task's from executing. Readiness might not be a boolean state though. Readiness may be more gradual as more side effects are produced. It is hard to configure more gradual startups where different features become ready at different times. The most generic way of specifying readiness is to allow a task to determine a stream or iterable of readiness states. And this stream would need to originate within the actual application itself and communicate with Nx. Even if readiness were simplified to be a boolean when all side effects are available, it would still have to be built into the application itself and cause delays in task execution.

#### Decision

For the initial implementation, option 1 of not introducing readiness state should suffice. Nx would still be able to add readiness handling in the future while remaining backwards compatible. Nx would default tasks to immediately become ready unless it has some other indicator for Nx to use. Readiness definitely needs to be handled but this will continue to stay within the tasks. As discussed before, both application servers and test tools already likely have some handling for this.

### Terminal Output

As discussed before in the different use cases, terminal output is a large part of how Nx would need to handle continuous tasks.

In general, terminal outputs of tasks can either be main or auxilliary. Main tasks are those directly requested by users and should be given prominent treatment. Tasks which main tasks depend on are auxilliary and can be optionally shown possibly at the same time. Nx does not currently have the infrastructure in place to support this.

However, utilizing Rust has already allowed Nx to be more transparent terminal output from individual tasks. In the coming year, we will be rewriting the orchestration of multiple task outputs via Rust. There are powerful terminal libraries in Rust that Nx could utilize to create a new terminal UI which can handle the above use cases.

This new UI would have 2 new capabilities that the current UI does not have:

#### Separate areas of terminal output

Nx should be able to output terminal output to separate areas of the terminal. This is necessary to view both main and auxilliary output at the same time. The auxilliary output should be handled differently depending on the width of the terminal. Terminals which are not wide enough will be limited to switching between main and auxilliary output.

#### Switch between outputs of different tasks

Multiple tasks can be main or auxilliary so Nx should allow for developers to switch between different tasks. This prevents the outputs from being interlaced which could make the output hard to understand.

### Current Terminal Output

The above terminal output will take some time to create so in the mean time, there still remains the existing display solutions for continuous tasks for the time being. When the new Terminal UI is ready, the developer experience using continuous tasks will improve. The following options would be switched by the developer with the current `--outputStyle` flag.

#### Option 1: Do not show terminal output for dependent tasks

In some situations, dependent tasks can very well be ignored. If all dependent tasks are working as they are expected to, then terminal outputs of dependent tasks could be hidden altogether. Nx already handles this today.

#### Option 2: Show terminal output interlaced with parent tasks

Interlacing terminal outputs is messy. Nx would prefix the outputs but either task's outputs could still be broken up in a way where it becomes hard to understand. In situations where a dependent task is not behaving as expected, this option can be used to debug the issue. This is not a great experience but Nx already currently has support for this.

## EDITS

I renamed the terminology to be "Continuous Tasks"

## Prior Art

https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/

Kubernetes allows containers to specify other containers to run alongside it much like Nx's tasks expect continuous tasks to run alongside them.

[Previous Sidecars RFC](https://github.com/nrwl/nx/discussions/23273)

This previous RFC designs out sidecars which are a subset of continuous tasks. In this design, continuous tasks depended upon by other tasks are synonymous to the sidecars mentioned in the previous design. This design encompasses more than just sidecar tasks.

**Comment by dwelch2344** [comment]
Yes please. Setting up things like SSH tunnels / etc would be so nice to do in a NX ecosystem

**Reply by leggomuhgreggo** [reply]
I'm not super familiar with the possible SSH tunnels use cases, but it strikes me as something that can already be handled using custom executors. 

To me, the key advantage of this RFC is that you can dynamically leverage the project graph to orchestrate tasks between different projects, ala dependsOn — but without violating boundaries — using an API that's still able to be declarative + modular. 

And it alleviates the need to for semi-omnipotent "god class" orchestrators, that knows too much about multiple projects

But, I donno, that could totally be a subjective take on my part. 

What do you think?

EDIT: After additional consideration, I can see "chained targets" in a single project as fair play. But to me the bigger unlock is still the modular, cross-project dynamic.

**Comment by lppedd** [comment]
Regarding the naming, I'd opt for something that is already widely used, such as

- Service task  

  For example, a Windows service starts up when the OS boots, and operates in background until explicitly terminated.
  Under Nx, explicit termination is termination of all Nx-related processes, or of all tasks that required the service task.

- Daemon task

  Same concept as a service, but it's probably more widely used because of Unix.

In the end we'll have
- Task - usually refers to a piece of work executed fairly quickly, part of a wider plan
- Long-running task - refers to a task that might take slightly longer to complete
- Daemon/Service task - refers to a long-running task that is explicitly marked as non-interactive, perpetually in background, and which allows other tasks to run concurrently

**Reply by FrozenPandaz** [reply]
Thank you for offering some alternative names! Service Task may make sense indeed.

I wouldn't use Daemon Task as Nx also has a daemon under the hood and it might be confused as a task that runs on the daemon :shrug:

When other tasks depend on an infinite task, it most probably is some sort of service indeed. However I would hesitate to call running `nx test --watch` starting a service.

Just my opinion :smile: Does calling these sorts of tasks services make sense to other people?

**Reply by lppedd** [reply]
Yeah it's difficult to pick a name that isn't ambiguous in all scenarios.  
The case of `test --watch` imo falls under the long-running task scenario. You're not starting a service task, but you're starting a task with an option that makes it behave as long-running.

**Reply by sceptre12** [reply]
Naming things are hard.. but I must ask.. from Nx's perspective.. Would the approach to your solution change when you're dealing with tasks that are long running, foreground tasks versus long running back ground task? 
I'm asking this because maybe the name for what your trying to solve can be scoped down to just solving one type of problem versus trying to create a catch all name for different types of "long-running tasks"

**Comment by beaussan** [comment]
Thanks a lot for this RFC, this is one of the most exciting changes I'm eager to use from Nx.

Long running task has been an issue for a while, especially with package.json#workspace resolution!

In terms of use cases, I'm very eager to use it for:
- depends on build watch of buildable libs
- openning ssh / tsh connexion to a backend
- starting the server / when staring the frontend, or a docker compose
- Start storybook before running storybook-test
- Running some codegen like `graphql-codegen --watch` as a watch infinite task

It opens soo many doors! :D

### One thought regarding `Readiness`

> Option 2: Add configuration
Tasks can report to Nx that the sidecar task produces a side effect. Nx can handle waiting for these side effects. Doing so would delay parent task's from executing. Readiness might not be a boolean state though. Readiness may be more gradual as more side effects are produced. It is hard to configure more gradual startups where different features become ready at different times. The most generic way of specifying readiness is to allow a task to determine a stream or iterable of readiness states. And this stream would need to originate within the actual application itself and communicate with Nx. Even if readiness were simplified to be a boolean when all side effects are available, it would still have to be built into the application itself and cause delays in task execution.

Nx already have a pattern regarding this, on the run-command [readyWhen](https://nx.dev/nx-api/nx/executors/run-commands#readywhen) option where a specific string on the output is used to mark a task as "done", but having the same pattern to mark a task as ready is not that far fetched.

While it's not critical to have it on the first deliverable, I do think it's super important, given docker start can take time for integration tests for example. The workaround would be to have the tasks that depends on a infinite tasks to `sleep 120` and hope it's not flaky

**Reply by yjaaidi** [reply]
I totally agree with Nicolas here about the importance of readiness and how it could start with something as minimalistic as `readyWhen`.

While I am an advocate of incremental development and pushing back features, I am worried that without readiness, infinite tasks adoption can be challenged.
In other words, while e2e test frameworks like Playwright provide solutions for that, most other tasks don't. This puts the responsability on users shoulders and can make the feature harder to adopt.

Without readiness, users might encounter some additional challenges like test flakiness. Here is an example:
- `e2e` depends on `frontend:serve`
- `frontend:serve` depends on `backend:serve`
- `backend:serve` depends on something else etc...

Users did a great job adding resilience to their app and retrying requests until the backend is up but considering the dependency chain, the test will start before all the apps and servers start causing flakiness and timeouts.
With something as primitive as `readyWhen`, users would be able to wait for servers to start before starting the app.

**Reply by aaronadamsCA** [reply]
Agreed readiness is very important, e.g. waiting for a port to open before continuing.

**Comment by sigerello** [comment]
> Nx should be able to output terminal output to separate areas of the terminal. This is necessary to view both main and auxilliary output at the same time.

Until it's rewritten in Rust, it can be relatively easy done with https://github.com/vadimdemedes/ink

**Comment by sceptre12** [comment]
I'm wondering if something special can be done to handle this type of situation:
### Multiple Unrelated Infinite Tasks

When working within a large monorepo and you have multiple systems running. Interlacing all of the terminal outputs at once can get very confusing and cumbersome. 
I'm wondering if its possible for ya'll to create a special flag for run-many that creates an interactive CLI that allows the user to view different streams of the log. 

I'm thinking of the current CLI that nx uses right. 
Say I execute something like this:

[nx run-many -t serve --cli --projects frontend,backend](https://nx.dev/nx-api/nx/documents/run-many#runmany)

The output provides a command interface which allows the user to select what stream they want to see. 
The user even has the ability to exit said stream and go and see another one. 
This would probably be way more complicated but you could make it so that the user can view multiple streams side by side. 

Just my 2cents but I think that would clean up all of the noise and allow the user to decided what they want to see when it comes to the terminal output

**Reply by romannurik** [reply]
This sounds really similar to [this thread I started about a possible Terminal UI](https://github.com/nrwl/nx/discussions/29098) ala turborepo. I also published a plugin ([nx-devui](https://github.com/romannurik/nx-devui//)) as an example of what it might look like.

**Reply by FrozenPandaz** [reply]
Yes, the UI we have in mind will allow uses to toggle between different tasks. 👍

**Comment by HaasStefan** [comment]
# Lifecycle hooks

I am curious if we can leverage the lifecycle hooks further and create a user API for it or some sort of notification stream.

I am currently on v19, using custom task runners for hand-rolled telemetry, collecting all kinds of data, such as the platform, task duration and much more. Due too the deprecation of custom task runners in v20 I am blocked from upgrading and there is no migration path for this use case yet. 

If you could produce a log with the lifecycles, or a pub/sub service, whatever, just something that I can use to trigger my custom telemetry service.

Do you think something like this could be possible?

**Comment by HaasStefan** [comment]
# Terminology

I like **persistent tasks** the most, as it is a common term in turborepo already and when someone new to monorepos would start to compare features, it would be easier to have common terminology among monorepo tools.

But I find **mortal tasks** for finite/short-lived tasks also really funny :)

But yeah, I think persistent and non-persistent is pretty straightforward.

**Comment by yjaaidi** [comment]
**Naming/ Terminology:** 1️⃣

Note: I am afraid that "immortal" might sound like a "daemon" process that Nx will keep up and running even after tasks finish.
_🤔 this could be an interesting feature by the way._

**Comment by yjaaidi** [comment]
I don't know if I am more excited about the new terminal UI or the infinite tasks 😊
This is all so exciting! Good job!

**Comment by yjaaidi** [comment]
Will infinite tasks be reused between tasks executed in different graphs?

Here is the use case:
- `app:e2e` task depends on infinite task `app:serve`
- user runs `nx serve app` in one terminal
- user runs `nx e2e app` in another terminal

If the Nx daemon could reuse the infinite tasks triggered in another graph:
- this would reduce the amount of work
- this would fix the common issue where users pass different options to the task to reuse existing servers etc...

**Reply by FrozenPandaz** [reply]
That would be neat.. It's not in scope for the initial implementation.. but it could be something that we add :thinking:

Could you create a separate feature request for that please? Once continuous tasks are stable we can revisit this.

**Comment by FrozenPandaz** [comment]
Hey all, I thought of a new name which I think is simpler but still specific.

**Continuous Tasks**

These tasks run continuously while other tasks run and until they are stopped.

What do you all think?

**Reply by leggomuhgreggo** [reply]
Nice! Yeah, I agree "continuous" sits easier than "infinite" (_infinity is spooky_)

If we're talking about the actual property name, my sense is that `detached: true` or `background: true` would be slightly more standard/customary — but `continuous: true` works too!

**Reply by moatorres** [reply]
I think **Continuous Tasks** is a great name. "Persistent Tasks" sounds related to persistence layers or something that will literally never end by means of self-restarting (even when arbitrarily shut down).

**Comment by leggomuhgreggo** [comment]
# Use Case: Federated `start` executor

## Workspace Context

Our company uses a Federated GraphQL API, where we have:
- a single **"router"** project (AKA "gateway"), using Apollo Router
- a number of **"subgraph"** service projects, using NestJS
- also some of our subgraphs have their own **"internal"** service dependencies

So the dir structure looks something like this:
```
├── apps
│   ├── router/
│   ├── subgraph-1/
│   ├── subgraph-2/
│   ├── subgraph-3/
│   └── interal-service-A/
```

### Subgraph Map

The `router` project has configuration file which maps all subgraphs to a URL

It looks something like this for local dev

```yaml
# apps/router/router.local.yml

subgraph-map:
  subgraph-1: http://localhost:1234/graphql
  subgraph-2: http://localhost:2345/graphql
  subgraph-3: http://localhost:3456/graphql
```  

And it looks sorta like this for development
```yaml
# apps/router/router.development.yml

subgraph-map:
  subgraph-1: https://subgraph-1.development.company.com/graphql
  subgraph-2: https://subgraph-2.development.company.com/graphql
  subgraph-3: https://subgraph-3.development.company.com/graphql
```  

## Objective

We want to be able to run `nx run router:start` and automatically spin up dev servers **for the affected subgraph projects**

Specifically:
- The subgraph map points to the remote/deployed development endpoints, by default
- If a developer has modified a particular subgraph project on their local branch, we want to:
  - start the local dev server for that project
  - update the router config with the new subgraph map
  - restart the router

So the subgraph map would have a mix or local and remote, like:
```yaml
# router.local.yml

subgraph-map:
  subgraph-1: http://localhost:1234/graphql <-- affected
  subgraph-2: https://subgraph-2.development.company.com/graphql
  subgraph-3: https://subgraph-3.development.company.com/graphql
```  

This way we only start local dev servers for the subgraph services that we need to.

## Problem

Ideally, we could use the existing `dependsOn` API to help determine which services need to be started.

This enables us to keep things dynamic — as opposed to hardcoding project details in some sort of over-powered (brittle) orchestration executor.

But there's currently not an easy way to have "detached" / "child" processes using `dependsOn`. (Hence this cool RFC!)


---

# Design Comments

I think the feature requirements have already been outlined fairly well in the RFC (nice!)

It would be awesome if we could leverage the existing `dependsOn` API — something along these lines

```json
  "targets": {
    "start": {
      "dependsOn": [
        {
          "target": "^start",
          "detached": true, 
        }
      ],
     ...
    },
```

### Possible Options

1. `"detached": true` - allows parent target to run, even if a given `dependsOn` process hasn't exited
1. `"cleanup": true`  - ends child process with parent SIGINT, SIGTERM (might always be true?)
1. `"readyWhen": "listening at "` - effectively a simple "wait on" (some tradeoffs here)
1. `"output-style": <output enum>` - I would think this API would translate fairly well, but haven't thought about it exhasutively

## Usage with Custom Executor?

I think it's fair to expect more advanced features (eg advanced wait on logic) to be handled by custom executors — and keep the target API features focused on the handling "detached" child processes.

To that end, there may be an argument for augmenting the executor context with child process IDs 

I know that the executor API is intentionally (and rightly) limited in scope. And I think it's important to maintain those design boundaries. (lifecycle hooks are dangerous!)

## Useful References (skippable)

I feel like the feature has a lot of conceptual overlap with the [concurrently](https://www.npmjs.com/package/concurrently) package — but I don't know if there's actually a whole lot of immediately compelling feature additions. 
Good reference point for semantics in any case. 

Hope that helps! 

Thanks!

**Comment by johnpyp** [comment]
> Continuous Tasks Depending on Discrete Tasks
> 
> Continuous tasks can depend on discrete tasks.
> 
> For example, app:serve could depend on itself and it's dependencies to be built before starting up.
> 
> Running a continuous task on its own with dependencies on other tasks but no dependencies on itself will be much like running a continuous task in isolation. The only difference is that Nx will wait for the discrete tasks to complete before the continuous task is started. **It is expected to continue running until Nx is interrupted (Ctrl + C).**

In Turborepo there is a concept of "interruptible" *and* persistent tasks. These are servers generally without hot reload or other *built-in* auto-restart mechanisms. Will the same be supported here?

**Comment by espipj** [comment]
A nice feature that I am not sure is possible to implement at the moment is to see the status of the continous task without needing to open the  terminal for that specific task.

Let say one of my tasks spins up a Express server that takes a while to load, some databases, some static generated stuff etc... and I run them as "run-many"
it would be great that when the Express server is ready an emoji shows up on that task or some status change, at the moment I can just see that it says "Continuous" and nothing else... So I need to open the terminal to check when it's ready

**Comment by LongLiveCHIEF** [comment]
The current implementation has some flaws that are making it difficult to actually work with in practice.  If I separate out my continuous tasks from their separate configurations into entirely new task targets, everything works... but it's certainly not efficient use of NX caching, and really feels like it breaks practice from nx's more "simple by default" approach that came with project crystal.

I wind up with a ton of `watch-something` tasks that have the same inputs/outputs and other configs as the `something` task.

If I mark a task as continuous, but then run a version of that task in a CI environment, even when the task succeeds it exits with a fail.

I think the approach of marking an entire task and all of its configurations as continuous, rather than being able to decorate a task configuration as 'continuous' violates the intent of the design principle behind nx.
